trin-core/src/portalnet/find/iterators/findnodes.rs

```diff
@@ -83,85 +89,76 @@ impl FindNodeQueryConfig {
+        self.started = Some(start);
+    }
+
+    fn on_failure(&mut self, peer: &TNodeId) {
+        if let QueryProgress::Finished = self.progress {
+            return;
+        }
+
+        let key: Key<TNodeId> = peer.clone().into();
+        let distance = key.distance(&self.target_key);
+
+        match self.closest_peers.entry(distance) {
+            Entry::Vacant(_) => {}
+            Entry::Occupied(mut e) => match e.get().state() {
+                QueryPeerState::Waiting(..) => {
+                    debug_assert!(self.num_waiting > 0);
+                    self.num_waiting -= 1;
+                    e.get_mut().set_state(QueryPeerState::Failed);
+                }
+                QueryPeerState::Unresponsive => e.get_mut().set_state(QueryPeerState::Failed),
+                _ => {}
+            },
         }
     }
 
-    /// Callback for delivering the result of a successful request to a peer
-    /// that the query is waiting on.
-    ///
     /// Delivering results of requests back to the query allows the query to make
     /// progress. The query is said to make progress either when the given
-    /// `closer_peers` contain a peer closer to the target than any peer seen so far,
+    /// `peer_response` contain a peer closer to the target than any peer seen so far,
     /// or when the query did not yet accumulate `num_results` closest peers and
-    /// `closer_peers` contains a new peer, regardless of its distance to the target.
-    ///
-    /// After calling this function, `next` should eventually be called again
-    /// to advance the state of the query.
-    ///
-    /// If the query is finished, the query is not currently waiting for a
-    /// result from `peer`, or a result for `peer` has already been reported,
-    /// calling this function has no effect.
-    pub fn on_success(&mut self, node_id: &TNodeId, closer_peers: Vec<TNodeId>) {
+    /// `peer_response` contains a new peer, regardless of its distance to the target.
+    fn on_success(&mut self, peer: &TNodeId, peer_response: Vec<TNodeId>) {
         if let QueryProgress::Finished = self.progress {
             return;
         }
 
-        let key: Key<TNodeId> = node_id.clone().into();
+        let key: Key<TNodeId> = peer.clone().into();
         let distance = key.distance(&self.target_key);
 
         // Mark the peer's progress, the total nodes it has returned and it's current iteration.
         // If the node returned peers, mark it as succeeded.
         match self.closest_peers.entry(distance) {
             Entry::Vacant(..) => return,
-            Entry::Occupied(mut entry) => match entry.get().state {
+            Entry::Occupied(mut entry) => match entry.get().state() {
                 QueryPeerState::Waiting(..) => {
                     assert!(
                         self.num_waiting > 0,
                         "Query has invalid number of waiting queries"
                     );
                     self.num_waiting -= 1;
```

Comment:
A `decrement_num_waiting` method might be nice to wrap up that check that's done multiple times. Though as far as I can see it would require the same implementation on both `Query` trait implementations, which wouldn't be very nice. I'm guessing that's why you didn't.

Reply:
Yeah, it could be a useful helper function, but I don't think it makes much of a difference either way. I will probably leave it for now.